### PR TITLE
update http options module

### LINKED ISSUE #12557: Add Confirmation Methods to HTTP Options
https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/auxiliary/scanner/http/options.md

https://nmap.org/nsedoc/scripts/http-methods.html

https://www.hackingarticles.in/multiple-ways-to-detect-http-options/

Also, the module should say (in verbose) that none were detected if none were detected.  Right now you just get no output at all.

### PR DESCRIPTION
**Pull request by h00die**
fix #12557

This PR fixes some of the issues with the http options module:
1. adds `targeturi` option as `/` isn't always the right answer
2. uses `send_request_cgi` because `raw` was just not needed
3. updated to http 1.1 instead of 1.0
4. lint
5. if a host doesn't give an `allow` header, print something so we have *some* feedback
6. attempt to handle the Tomcat error page

I was unable to find a host that had the Tomcat error page, but there are a bunch of internet references to it, so I did my best to code it in.

## Verification

- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/http/options`
- [x] `set rhosts`
- [x] `set verbose true`
- [x] `run`
- [x] **Verify** if the host doesn't respond with `Allow` header, you get an error message
- [x] **Verify** hosts with `Allow` headers still work

# old output

```
msf6 auxiliary(scanner/http/options) > run

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

# new output

```
msf6 auxiliary(scanner/http/options) > run

[-] 1.1.1.1 missing Allow header
[*] Scanned 1 of 1 hosts (100% complete)
```

### CODE REVIEW

File: modules/auxiliary/scanner/http/options.rb [hunk 1]
```diff
@@ -5,69 +5,80 @@
...
+        OptString.new('TARGETURI', [true, 'URI to test', '/']),
       ]
     )
   end
 
   def run_host(target_host)
+    res = send_request_cgi({
+      'uri' => datastore['TARGETURI'],
+      'method' => 'OPTIONS'
+    }, 10)
```
Comment (adfoster-r7) [inline modules/auxiliary/scanner/http/options.rb:40]:
```suggestion
    })
```

I know this was an existing value; but I'm not sure if the hard coded timeout was intentional here 👀

File: modules/auxiliary/scanner/http/options.rb [hunk 2]
```diff
@@ -5,69 +5,80 @@
...
+    res = send_request_cgi({
+      'uri' => datastore['TARGETURI'],
+      'method' => 'OPTIONS'
+    })
+    return if !res
+
+    # patch so that tomcat which responds in HTTP will work as well
+    # https://github.com/rapid7/metasploit-framework/issues/12557#issuecomment-552263162
+    # https://stackoverflow.com/questions/23886941/http-status-405-jsps-only-permit-get-post-or-head
+    if res.body =~ %r{<h1>HTTP Status 405 - JSPs only permit ([^<]*)</h1>}
```
Comment (gwillcox-r7) [inline modules/auxiliary/scanner/http/options.rb:46]: Not sure we necessarily need to match on these H1 tags since that might change, what are your feelings on removing them from the regex?

File: modules/auxiliary/scanner/http/options.rb [hunk 3]
```diff
@@ -5,69 +5,80 @@
...
 
-        report_note(
-          :host	=> target_host,
-          :proto => 'tcp',
-          :sname => (ssl ? 'https' : 'http'),
-          :port	=> rport,
-          :type	=> 'HTTP_OPTIONS',
-          :data	=> res.headers['Allow']
-        )
+    print_good("#{target_host} allows #{allowed_methods} methods")
```
Comment (gwillcox-r7) [inline modules/auxiliary/scanner/http/options.rb:57]: Is this not an array of items normally? Your edge case seems to be converting this into a string but I'm not sure this code will work in normal circumstances.

**Review by gwillcox-r7 — CHANGES_REQUESTED** [review]

File: modules/auxiliary/scanner/http/options.rb [hunk 4]
```diff
@@ -5,69 +5,80 @@
 
 class MetasploitModule < Msf::Auxiliary
 
-  # Exploit mixins should be called first
   include Msf::Exploit::Remote::HttpClient
   include Msf::Auxiliary::WmapScanServer
-  # Scanner mixin should be near last
```
[the remark this answers is not in the document]
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:11]: they felt very generic and like things we'd put in an example file for people to learn with.  Just generally not needed.

File: modules/auxiliary/scanner/http/options.rb [hunk 5]
```diff
@@ -5,69 +5,80 @@
...
+    end
 
-    begin
-      res = send_request_raw({
-        'version'      => '1.0',
-        'uri'          => '/',
-        'method'       => 'OPTIONS'
-      }, 10)
+    unless res.headers['Allow']
+      vprint_error("#{target_host} missing Allow header")
```
[the remark this answers is not in the document]
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:51]: when scanning a large address space (class C), I would by default think you'd only want to see alive valid things, and verbosely see things that aren't there

File: modules/auxiliary/scanner/http/options.rb [hunk 6]
```diff
@@ -5,69 +5,80 @@
...
   end
 
   def run_host(target_host)
+    res = send_request_cgi({
+      'uri' => datastore['TARGETURI'],
+      'method' => 'OPTIONS'
+    })
+    return if !res
+
+    # patch so that tomcat which responds in HTTP will work as well
```
[the remark this answers is not in the document]
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:43]: I've re-worded this to be more clear

File: modules/auxiliary/scanner/http/options.rb (hunk 2 above)
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:46]: It may change, but I think this is also an edge case for a likely outdated version as I couldn't find any in the wild to test it against, so I think its safe to keep at this point.

File: modules/auxiliary/scanner/http/options.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/scanner/http/options.rb:11]: My concern was that these were added to prevent issues with people adding dependencies as includes and then breaking things as a result of bad ordering. Granted this is a module and not a library so perhaps this is less needed since its less likely to break a whole suite of things, but I still think its good practice to document cases where things might fail if something was to be changed in a way that would be expected to work but in reality won't work.

File: modules/auxiliary/scanner/http/options.rb (hunk 3 above)
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:60]: during testing I also added `puts res.headers['Allow'].class` and received `String` in the output, so I believe everything is a string that I'm dealing with.

File: modules/auxiliary/scanner/http/options.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:11]: i'll add them back, nbd

File: modules/auxiliary/scanner/http/options.rb (hunk 3 above)
Comment (gwillcox-r7) [reply modules/auxiliary/scanner/http/options.rb:62]: Thanks that is sufficient for me, will mark this as resolved, don't think this should be an issue.

File: modules/auxiliary/scanner/http/options.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/scanner/http/options.rb:56]: The caveat here is what if you are scanning just a single host? It will fail silently.

File: modules/auxiliary/scanner/http/options.rb [hunk 7]
```diff
@@ -14,60 +14,76 @@ class MetasploitModule < Msf::Auxiliary
...
     )
   end
 
   def run_host(target_host)
+    res = send_request_cgi({
+      'uri' => datastore['TARGETURI'],
+      'method' => 'OPTIONS'
+    })
+    return unless res
+    return if res.nil?
```
Comment (gwillcox-r7) [inline modules/auxiliary/scanner/http/options.rb:44]: These two lines are pretty much equivalent? I imagine we can skip the `.nil? case and keep the first line.

File: modules/auxiliary/scanner/http/options.rb [hunk 8]
```diff
@@ -5,69 +5,80 @@
...
       ]
     )
   end
 
   def run_host(target_host)
+    res = send_request_cgi({
+      'uri' => datastore['TARGETURI'],
+      'method' => 'OPTIONS'
+    })
+    return if !res
```
[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/scanner/http/options.rb]: Updates look fine, marking as resolved.

File: modules/auxiliary/scanner/http/options.rb (hunk 2 above)
Comment (gwillcox-r7) [reply modules/auxiliary/scanner/http/options.rb]: Hmm fair point, lets keep it in as is, and we can update it later if needed.

File: modules/auxiliary/scanner/http/options.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (h00die) [reply modules/auxiliary/scanner/http/options.rb:56]: Being an aux scanner module, I assume its mostly used for large block scans.  Also per my original issue:
`Also, the module should say (in verbose) that none were detected if none were detected. Right now you just get no output at all.`
So the original module, no matter what, gave no output for a host w/o an Allow header.  This at least gives you an option to see it if need be

[the remark this answers is not in the document]
Comment (gwillcox-r7) [reply modules/auxiliary/scanner/http/options.rb:56]: Fair point and after reviewing this further I came to the same conclusion r.e the code in that a bigger change would be needed to incorporate what I was suggesting. I'll leave this as is, your fix is in line with existing behavior 👍

**Comment by gwillcox-r7** [comment]
Everything looks good, will fix up the one minor remaining comment and then test and land this 👍

**Comment by gwillcox-r7** [comment]
Looks like baseline still works, now to test a host that doesn't respond with an Allow header:

```
msf6 > use auxiliary/scanner/http/options 
msf6 auxiliary(scanner/http/options) > set RHOSTS www.google.com
RHOSTS => www.google.com
msf6 auxiliary(scanner/http/options) > set VERBOSE true
VERBOSE => true
msf6 auxiliary(scanner/http/options) > run

[+] 142.250.138.105 allows GET, HEAD methods
[*] Scanned 1 of 7 hosts (14% complete)
[+] 142.250.138.99 allows GET, HEAD methods
[*] Scanned 2 of 7 hosts (28% complete)
[+] 142.250.138.147 allows GET, HEAD methods
[*] Scanned 3 of 7 hosts (42% complete)
[+] 142.250.138.104 allows GET, HEAD methods
[*] Scanned 4 of 7 hosts (57% complete)
[+] 142.250.138.103 allows GET, HEAD methods
[*] Scanned 5 of 7 hosts (71% complete)
[+] 142.250.138.106 allows GET, HEAD methods
[*] Scanned 6 of 7 hosts (85% complete)
[-] The host ([2607:f8b0:4000:81b::2004]:80) was unreachable.
[*] Scanned 7 of 7 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/http/options) >
```

And on a host without the allow header things are working as expected:

```
msf6 auxiliary(scanner/http/options) > show options

Module options (auxiliary/scanner/http/options):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS     www.rapid7.com   yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT      80               yes       The target port (TCP)
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       URI to test
   THREADS    1                yes       The number of concurrent threads (max one per host)
   VHOST                       no        HTTP server virtual host


View the full module info with the info, or info -d command.

msf6 auxiliary(scanner/http/options) > set RPORT 443
RPORT => 443
msf6 auxiliary(scanner/http/options) > run

[-] 13.249.74.44 missing Allow header
[*] Scanned 1 of 4 hosts (25% complete)
[-] 13.249.74.39 missing Allow header
[*] Scanned 2 of 4 hosts (50% complete)
[-] 13.249.74.83 missing Allow header
[*] Scanned 3 of 4 hosts (75% complete)
[-] 13.249.74.60 missing Allow header
[*] Scanned 4 of 4 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/http/options) > set SSL true
[!] Changing the SSL option's value may require changing RPORT!
SSL => true
msf6 auxiliary(scanner/http/options) > run

[-] 13.249.74.83 missing Allow header
[*] Scanned 1 of 4 hosts (25% complete)
[-] 13.249.74.44 missing Allow header
[*] Scanned 2 of 4 hosts (50% complete)
[-] 13.249.74.60 missing Allow header
[*] Scanned 3 of 4 hosts (75% complete)
[-] 13.249.74.39 missing Allow header
[*] Scanned 4 of 4 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/http/options) > 
```

**Comment by gwillcox-r7** [comment]
LGTM will land this once tests pass!

**Comment by gwillcox-r7** [comment]
# Release Notes
An update has been made to the `modules/auxiliary/scanner/http/options.rb` module to modernize a few of its options, tidy up the code, and to handle an edge case when a target server might respond with a Tomcat error page.

**Review by gwillcox-r7 — APPROVED** [review]

**Comment by h00die** [comment]
thanks for the cleanup!
